File: TestCode/golang/c/ArrayParameter.go
package main

import (
  "fmt"
)
func sum( ap []int, n int) int {
  var m int
  for i := 0; i < 10; i++ {
    m += ap[i]
  }
  return m
}
func change(a []int, n int)  {
  a[n] = 900
  fmt.Printf("change : &n=%p,&a=%p,a=%p\n", &n, &a, a)
  fmt.Printf("change a[%d] = %d\n", n, a[n])
}
func main()  {
//  var  a [10]int = [10]int {1,2,3,4,5,6,7,8,9,10}
  a := []int {1,2,3,4,5,6,7,8,9,10}
  fmt.Printf("sum = %d\n", sum(a, 10))

  n := 5
  fmt.Printf("a[%d] = %d\n", n, a[n])
  fmt.Printf("Before : &n=%p,&a=%p,a=%p\n", &n, &a, a)
  change(a, n)
  fmt.Printf("After  : &n=%p,&a=%p,a=%p\n", &n, &a, a)
  fmt.Printf("after a[%d] = %d\n", n, a[n])
}

File: TestCode/golang/c/ArrayParameter.cpp
#include <iostream>
#include <stdio.h>

int sum (int ap[], int n){
  int m = 0;
  for (int i=0; i < n;  i++ ) {
    /* code */
    m += * ap;
    ap++;
  }
  return m;
}
void change(int a[], int n){
  a[n] = 900;
  fprintf(stdout,"change : &n =%p,&a=%p, a=%p\n",&n,&a,a);
}
int main(){
  int a[10] ={1,2,3,4,5,6,7,8,9,10};
  std::cout << "sum = " << sum(a, 10) << '\n';
  int n = 5;
  std::cout << "a[" << n << "] = " << a[n] << '\n';
  fprintf(stdout,"Before : &n =%p,&a=%p, a=%p\n",&n,&a,a);
  change(a, n);
  fprintf(stdout,"After  : &n =%p,&a=%p, a=%p\n",&n,&a,a);
  std::cout << "after a[" << n << "] = " << a[n] << '\n';
  return 0;
}
